### PR TITLE
Swap closure state in the type-checker to `var` to avoid TDZ checks.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Follow-up from #52832.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this faster

**Comment by jakebailey** [comment]
It's sort of interesting that it's always 0.10ish seconds saved; is there a fixed window where this is problematic?

**Comment by DanielRosenwasser** [comment]
Ugh, probably conflating changes.

@typescript-bot perf test this

**Comment by jakebailey** [comment]
The parser scanner change was a huge win for the few variables it modified, but this one is 10x less of a win for like 4x the variables changed (a few hundred); do we think it's worth it?

**Comment by jakebailey** [comment]
Well, I guess xstate is consistently like 3%.

**Comment by DanielRosenwasser** [comment]
> do we think it's worth it?

Definitely! I think it's very scoped, and the conceptual cleanliness buys us relatively little - so it's a free win to me. I could never argue with a developer using TypeScript that a 1-3% reduction in type-checking time wasn't worth it because we didn't want to disable our linter.

So I think we should even do this across the compiler. The only thing that might be lacking is some documentation on why it is that we're doing this.

**Review by jakebailey — APPROVED** [review]
I still don't like it given how it's like 300 variables, but I guess LGTM given they're at the top and rarely touched.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1378,12 +1378,13 @@ export function isInstantiatedModule(node: ModuleDeclaration, preserveConstEnums
...
 export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
-    const getPackagesMap = memoize(() => {
+    /* eslint-disable no-var */
+    var getPackagesMap = memoize(() => {
         // A package name maps to true when we detect it has .d.ts files.
         // This is useful as an approximation of whether a package bundles its own types.
         // Note: we only look at files already found by module resolution,
         // so there may be files we did not consider.
-        const map = new Map<string, boolean>();
+        var map = new Map<string, boolean>();
```
[the remark this answers is not in the document]
Comment (SlurpTheo) [reply src/compiler/checker.ts:1387]: Wait, why did you suggest reverting just this one back to `const`

[the remark this answers is not in the document]
Comment (jakebailey) [reply src/compiler/checker.ts:1387]: It's not a top level variable and is only referenced once for this memoized call. There's no way it has any meaningful performance impact.

But, my suggestion wasn't merged anyway and it's not hurting anything either.

[the remark this answers is not in the document]
Comment (SlurpTheo) [reply src/compiler/checker.ts:1387]: Maybe/probably because it's not "top level"?

**Pull request merged**